webpack/assets/javascripts/react_app/common/MountingService.js

```diff
@@ -7,21 +7,49 @@ import componentRegistry from '../components/componentRegistry';
 
 export { default as registerReducer } from '../redux/reducers/registerReducer';
 
-const mountedNodes = [];
+/**
+ * mounted-nodes cache
+ */
+let mountedNodes = [];
 
 // In order to support turbolinks with react, need to
 // unmount all root components before turbolinks do the unload
 // TODO: remove it when migrating into (webpacker-react or react-rails)
-document.addEventListener('page:before-unload', () => {
-  let node = mountedNodes.shift();
+document.addEventListener('page:before-unload', () => unmountComponents());
+
+const unmountComponentsByNodes = (...nodes) => {
+  let node = nodes.shift();
 
   while (node) {
     ReactDOM.unmountComponentAtNode(node);
-    node = mountedNodes.shift();
+    node = nodes.shift();
   }
-});
+};
+
+/**
+ * Will unmount all react-components from the dom
+ * @param  {string} selector unmount component inside the selector
+ * @return {number}          amount of unmounted nodes
+ */
+export const unmountComponents = (selector = null) => {
+  let nodesToUnmount = mountedNodes;
+  let nodesToKeep = [];
+
+  if (selector) {
+    const reactNode = document.querySelector(selector);
+    const shouldUnmount = node => reactNode.contains(node);
+
+    nodesToUnmount = mountedNodes.filter(node => shouldUnmount(node));
+    nodesToKeep = mountedNodes.filter(node => !shouldUnmount(node));
```

Comment:
Would it make sense to use some lodash function (or some other magic) here to calculate one of the arrays as the difference between the original array and the other one, instead of checking all of the nodes against the dom?

Reply:
I actually tried to go this way but found my code long and not that readable.

Do you think it's that bad?
I personally prefer to keep it readable then well-performance because:
1. Right now, on average we should have 4-6 items in this array.
2. It's a temporary solution to support turbolinks and we will eventually rid of it

Reply:
My main concern is that if this list gets long it could be slow - each node on the list calls a recursive check on the subtree of the parent node. for example, on the host index page this might be tens of elements (power status per line). If it makes the code significantly more complex then I don't mind leaving as is and only optimizing if we see actual performance impact.